include/benchmark/benchmark_api.h

```diff
@@ -239,6 +240,149 @@ inline BENCHMARK_ALWAYS_INLINE void DoNotOptimize(Tp const& value) {
 // FIXME Add ClobberMemory() for non-gnu compilers
 #endif
 
+
+
+class Counter {
+public:
+
+  typedef enum {
+    kDefaults       = 0x00,
+    /** Mark the counter as a rate. It will be presented divided by the duration of the benchmark. */
+    kIsRate         = 0x01 << 0,
+    /** Mark the counter as a thread-average quantity. It will be presented divided by the number of threads. */
+    kAvgThreads     = 0x01 << 1
+  } Flags;
+
+public:
+
+  Counter();
+  Counter(const char *name);
+  Counter(const char *name, double v);
+  Counter(const char *name, double v, uint32_t flags);
+  ~Counter();
+  Counter(Counter const& that);
+  Counter& operator= (Counter const& that);
+#ifdef BENCHMARK_HAS_CXX11
+  Counter(Counter && that);
+  Counter& operator= (Counter && that);
+#endif
+
+  void Set(double v) { value_ = v; }
+
+  void Finish(double cpu_time, double num_threads);
+
+  const char* Name() const { return name_; }
+  double Value() const { return value_; }
+
+  bool SameName(const char *n) const;
+
+  Counter& operator = (double v) { value_  = v; return *this; }
+  Counter& operator+= (double v) { value_ += v; return *this; }
+  Counter& operator-= (double v) { value_ -= v; return *this; }
+  Counter& operator*= (double v) { value_ *= v; return *this; }
+  Counter& operator/= (double v) { value_ /= v; return *this; }
+
+private:
+
+  // using raw C strings here to avoid inclusion of std::string in the public api
+  char*    name_;     // this points to either name_buf_ or name_mem_
+  double   value_;
+  uint32_t flags_;
+
+  char name_buf_[16]; // names smaller than 15 characters are stored here
+  char* name_mem_;    // whereas longer strings are stored here
+  size_t mem_size_;   // current size of name_mem_
+
+  void SetName_(const char *name);
+};
+
+
+
+class BenchmarkCounters {
+public:
+
+  typedef Counter        value_type;
+  typedef Counter const* const_iterator;
+
+public:
+
+  BenchmarkCounters();
+  ~BenchmarkCounters();
+  BenchmarkCounters(BenchmarkCounters const& that);
+  BenchmarkCounters& operator= (BenchmarkCounters const& that);
+#ifdef BENCHMARK_HAS_CXX11
+  BenchmarkCounters(BenchmarkCounters && that);
+  BenchmarkCounters& operator= (BenchmarkCounters && that);
+#endif
+
+  // Get a counter by name. Complexity: O(nameLength*size())
+  Counter const& operator[] (const char *name) const;
+  // Get (or insert) a counter by name. Complexity: O(nameLength*size())
+  Counter & operator[] (const char *name);
+
+
+  // Get a counter by id. Complexity: O(1)
+  Counter & operator[] (size_t id);
+  // Get a counter by id. Complexity: O(1)
+  Counter const& operator[] (size_t id) const;
+
+
+  // Insert a counter. Return its id. Complexity: O(nameLength*size())
+  size_t Insert(const char* name);
+  // Insert or update a counter. Return its id. Complexity: O(nameLength*size())
+  size_t Insert(const char *name, double value);
```

Comment:
It would be nice if the style was consistent, eg with the way that `*` and `&` are glued to the type or the variable (compare the two previous declarations).  I would suggest to run clang-format with style Google.